MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs

```diff
@@ -4954,56 +4978,147 @@ public string GetHwTonemapFilter(EncodingOptions options, string hwTonemapSuffix
+            EncodingOptions options,
+            string vidDecoder,
+            string vidEncoder)
+        {
             var inW = state.VideoStream?.Width;
             var inH = state.VideoStream?.Height;
             var reqW = state.BaseRequest.Width;
             var reqH = state.BaseRequest.Height;
             var reqMaxW = state.BaseRequest.MaxWidth;
             var reqMaxH = state.BaseRequest.MaxHeight;
             var threeDFormat = state.MediaSource.Video3DFormat;
-            var newfilters = new List<string>();
-            var noOverlay = swFilterChain.OverlayFilters.Count == 0;
-            var supportsHwDeint = _mediaEncoder.SupportsFilter("yadif_videotoolbox");
-            // fallback to software filters if we are using filters not supported by hardware yet.
-            var useHardwareFilters = noOverlay && (!doDeintH2645 || supportsHwDeint);
 
-            if (!useHardwareFilters)
+            var isVtEncoder = vidEncoder.Contains("videotoolbox", StringComparison.OrdinalIgnoreCase);
+            var hwScaleFilter = GetHwScaleFilter("vt", null, inW, inH, reqW, reqH, reqMaxW, reqMaxH);
+
+            var doDeintH264 = state.DeInterlace("h264", true) || state.DeInterlace("avc", true);
+            var doDeintHevc = state.DeInterlace("h265", true) || state.DeInterlace("hevc", true);
+            var doDeintH2645 = doDeintH264 || doDeintHevc;
+            var doVtTonemap = IsVideoToolboxTonemapAvailable(state, options);
+            var doOclTonemap = !doVtTonemap && IsHwTonemapAvailable(state, options);
+            var doTonemap = doVtTonemap || doOclTonemap;
+
+            var doScale = !string.IsNullOrEmpty(hwScaleFilter);
+            var hasSubs = state.SubtitleStream is not null && state.SubtitleDeliveryMethod == SubtitleDeliveryMethod.Encode;
+            var hasTextSubs = hasSubs && state.SubtitleStream.IsTextSubtitleStream;
+            var hasGraphicalSubs = hasSubs && !state.SubtitleStream.IsTextSubtitleStream;
+            var hasAssSubs = hasSubs
+                && (string.Equals(state.SubtitleStream.Codec, "ass", StringComparison.OrdinalIgnoreCase)
+                    || string.Equals(state.SubtitleStream.Codec, "ssa", StringComparison.OrdinalIgnoreCase));
+
+            // FIXME: scale_vt lacks of format option for the time being.
+            // hwdownload/hwmap to sw requires setting a format explicitly.
+            if (!isVtEncoder)
             {
-                return swFilterChain;
+                // should not happen.
+                return (null, null, null);
             }
 
-            // ffmpeg cannot use videotoolbox to scale
-            var swScaleFilter = GetSwScaleFilter(state, options, vidEncoder, inW, inH, threeDFormat, reqW, reqH, reqMaxW, reqMaxH);
-            newfilters.Add(swScaleFilter);
+            /* Make main filters for video stream */
+            var mainFilters = new List<string>();
+
+            if (!(doTonemap || doScale || hasSubs || doDeintH2645))
```

Comment:
I don't get it, why would you check these prematurely? Subsequent code has checked them and will eliminate empty filters.

When we enable more formats for the VT decoder in the future, or when upstream adds new ones to the format list of the VT encoder, the sw formats in this code path will not be subject to these known-working formats and may throw runtime error.

I think what we want is some concrete format rather than letting ffmpeg negotiate it.

```cs
            mainFilters.Add("format=nv12|p010le|videotoolbox_vld");
            mainFilters.Add("hwupload=derive_device=videotoolbox");
```

Reply:
This is because I've seen empty hwupload causing out-of-order frames during my early testing, but I'm not seeing them again since we stopped using hw surface. So, yes, perhaps we don't need to check this prematurely for now.